Add import progress report

NeimadTL opened the pull request:
Related issue: [7220](https://github.com/ydb-platform/ydb/issues/7220)

Although the progress report is present for the export operation, it was missing for the import. It's been added for both for feature consistency and for allowing users to know how long the operation is going to take.

NeimadTL committed:
Add import progress report

Although the progress report is present for the export operation, it was missing
for the import. It's added for both for feature consistency and for allowing users
to know how long the operation is going to take.

NeimadTL commented:
@pixcc @maximyurchuk , is that something [that](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/PR-check/12127201720/ya-x86-64/ya_make_output.txt) that should be fixed on my end ?

FYI, I did run the `ya test` command but it was taking so long that stopped and went ahead and pushed.

NeimadTL committed:
Fix typo

The argument was mispelled and therefore an error was thrown.

pixcc commented:
@NeimadTL Thank you for your contribution! Going to review it as fast as possible

NeimadTL commented:
> @NeimadTL Thank you for your contribution! Going to review it as fast as possible.

Awesome ! you're welcome.

NeimadTL committed:
Fix typo and change id used for importpath

The `TImportInfo` does not have a `importPathId` member in contrary to
the `TExportInfo`. My guess is that the `DomainPathId` my be the one used for
building the import path.

NeimadTL commented:
@pixcc, when reviewing please take into account my last commit's message. If my guess is wrong, the other suggestion I have is to do `TPathId ImportPathId = InvalidPathId;` on the `TImportInfo` struc defined in `ydb/core/tx/schemeshard/schemeshard_info_types.h #L2776`

NeimadTL committed:
Include missing header and iterate over import items

A header was forgotten and therefore `ImportItemPathName` was
unknown. Also, a for loop has been added to iterate over each import
items.

NeimadTL committed:
Fix test case

The permissions were missing and the test was setup for export instead
of import.

NeimadTL commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -28,6 +29,48 @@ namespace {
+            itemProgress.set_parts_completed(txState.Shards.size() - txState.ShardsInProgress.size());
+            *itemProgress.mutable_start_time() = SecondsToProtoTimeStamp(txState.StartTime.Seconds());
+        } else {
+            const auto path = TPath::Resolve(ImportItemPathName(ss, importInfo, itemIdx), ss);
+            if (!path.IsResolved() || !ss->Tables.contains(path.Base()->PathId)) {
```
Thanks for your feedback. 

So basically, what you're is that L49 to L52 are not needed, correct ? and then, on L54 I do the following:
`auto table = ss->Tables.at(path.Base()->DstPathId);`

pixcc replied:
L49 is not needed. L50 is partially needed, you need to check that `Tables` contains `DstPathId`.

> `auto table = ss->Tables.at(path.Base()->DstPathId);` ?

almost, actually this `DstPathId` is stored in `TItem` struct.
https://github.com/ydb-platform/ydb/blob/fb2c723ca8e9b4b49f492db5625af6b897327de9/ydb/core/tx/schemeshard/schemeshard_info_types.h#L2819

You can get this `TItem` object from `items` vector (dynamic array) stored in `importInfo`
https://github.com/ydb-platform/ydb/blob/fb2c723ca8e9b4b49f492db5625af6b897327de9/ydb/core/tx/schemeshard/schemeshard_info_types.h#L2857

NeimadTL replied:
Thanks @pixcc. Just pushed again.

NeimadTL commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4006,6 +4006,86 @@ Y_UNIT_TEST_SUITE(TImportTests) {
+        )");
+        env.TestWaitNotification(runtime, txId);
+
+        const auto permissions = R"(
```
Thanks

NeimadTL commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4006,6 +4006,86 @@ Y_UNIT_TEST_SUITE(TImportTests) {
+        TS3Mock s3Mock(ConvertTestData(data), TS3Mock::TSettings(port));
+        UNIT_ASSERT(s3Mock.Start());
+
+        const TString userSID = "user@builtin";
```
Ok

NeimadTL commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4006,6 +4006,86 @@ Y_UNIT_TEST_SUITE(TImportTests) {
+
+        const TString userSID = "user@builtin";
+        TestImport(runtime, ++txId, "/MyRoot", Sprintf(R"(
+            ExportToS3Settings {
```
Thank you for that.

NeimadTL committed:
Fix test & make use of import data instead export

Permission and user were not needed in the test so they've been removed.
The import FillItemProgress used export data instead of import data so
it's updated accordingly.

Also, in export case tables (items) are copied in temporary folder inside
database. If one tries to export table `MyTable`, it is copied to a temporary
folder created specifically for this export. In this folder `MyTable` is
named by its itemIdx.

├── MyTable
│
└── export-temporary-folder
    │
    └── 0 // MyTable copy

After that, data is exported from this copied table. This copy operation is
required to reduce the load on tables during exporting. In case of import,
there is no temporary folder. Data is imported directly in destination tables
and that's why the `DstPathId` field has been used of item instead of creating
path name, resolving it and getting its id on L49 and L53.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import_flow_proposals.cpp:
```diff
@@ -226,5 +226,14 @@ THolder<TEvIndexBuilder::TEvCancelRequest> CancelIndexBuildPropose(
     return MakeHolder<TEvIndexBuilder::TEvCancelRequest>(ui64(indexBuildId), domainPath.PathString(), ui64(indexBuildId));
 }
 
+TString ImportItemPathName(TSchemeShard* ss, const TImportInfo::TPtr importInfo, ui32 itemIdx) {
```
unused now

NeimadTL replied:
good catch.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -1,6 +1,7 @@
 #include "schemeshard_import.h"
 #include "schemeshard_import_helpers.h"
 #include "schemeshard_impl.h"
+#include "schemeshard_import_flow_proposals.h"
```
don't need now

NeimadTL replied:
good catch.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -28,6 +29,47 @@ namespace {
+                return;
+            }
+
+            auto table = ss->Tables.at(item->DstPathId);
```
`item.DstPathId`

`->` notation is used, when you accessing field through pointer

NeimadTL replied:
Thanks for that.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -28,6 +29,47 @@ namespace {
+            itemProgress.set_parts_completed(txState.Shards.size() - txState.ShardsInProgress.size());
+            *itemProgress.mutable_start_time() = SecondsToProtoTimeStamp(txState.StartTime.Seconds());
+        } else {
+            if (!ss->Tables.contains(item->DstPathId)) {
```
`item.DstPathId`

`->` notation is used, when you accessing field through pointer

NeimadTL commented on ydb/core/tx/schemeshard/schemeshard_import_flow_proposals.h:
```diff
@@ -46,5 +46,8 @@ THolder<TEvIndexBuilder::TEvCancelRequest> CancelIndexBuildPropose(
 );
 
+TString ImportItemPathName(TSchemeShard* ss, const TImportInfo::TPtr importInfo, ui32 itemIdx);
+TString ImportItemPathName(const TString& importPathName, ui32 itemIdx);
```
good catch.

NeimadTL committed:
Remove unsed references

The ImportItemPathName was not needed but its reference were still
present. Also, as `item` is not a pointer, the call should be made with
`.` instead of `->`.

NeimadTL commented:
@pixcc, just pushed the requested changes. Thanks again for your help. Hopefully this time around, it'll be good.

NeimadTL committed:
Remove forgotten reference

pixcc commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4006,6 +4006,61 @@ Y_UNIT_TEST_SUITE(TImportTests) {
+              type { optional_type { item { type_id: UTF8 } } }
+            }
+            primary_key: "key"
+        )", {{"a", 1}}, permissions);
```
Forgot to remove `permissions` here. Tests won't compile :(

NeimadTL replied:
thanks

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import_flow_proposals.cpp:
```diff
@@ -226,5 +226,6 @@ THolder<TEvIndexBuilder::TEvCancelRequest> CancelIndexBuildPropose(
     return MakeHolder<TEvIndexBuilder::TEvCancelRequest>(ui64(indexBuildId), domainPath.PathString(), ui64(indexBuildId));
 }
 
+
```
Since you still have to do another commit for the tests to compile. Could you remove the extra newline here?

NeimadTL replied:
yes

pixcc commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4006,6 +4006,61 @@ Y_UNIT_TEST_SUITE(TImportTests) {
             NLs::HasNoRight("+R:bob")
         });
     }
+
```
Since you still have to do another commit for the tests to compile. Could you remove the extra newline here?

pixcc commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4006,6 +4006,61 @@ Y_UNIT_TEST_SUITE(TImportTests) {
+        TTestEnv env(runtime);
+        ui64 txId = 100;
+
+        TestCreateTable(runtime, ++txId, "/MyRoot", R"(
```
Looks like you don't need it too

NeimadTL replied:
ok, thanks

NeimadTL replied:
appreciate it

NeimadTL committed:
Remove unnecessary code and extra blank lines

Unsed were still present and extra blank line as well so both have been
removed.

NeimadTL commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -67,7 +108,9 @@ void TSchemeShard::FromXxportInfo(NKikimrImport::TImport& import, const TImportI
             break;
         case TImportInfo::EState::Transferring:
-            // TODO(ilnaz): fill items progress
+            for (ui32 itemIdx : xrange(importInfo->Items.size())) {
```
well, I had seen in the ternary operator in the export and but didn't do in the import as it's already taken care of on [L85](https://github.com/ydb-platform/ydb/blob/main/ydb/core/tx/schemeshard/schemeshard_import.cpp#L85). What do you think ?

NeimadTL committed:
Add progress fpr `DONE` case

In order to be consistent with the export operation, the progression has
been added for the import's `DONE` case as well.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -67,15 +108,17 @@ void TSchemeShard::FromXxportInfo(NKikimrImport::TImport& import, const TImportI
+            for (ui32 itemIdx : xrange(importInfo->Items.size())) {
+                FillItemProgress(this, importInfo, itemIdx, *import.AddItemsProgress());
+            }
+            import.SetProgress(importInfo->IsDone()
```
Build is failed

```
/home/runner/actions_runner/_work/ydb/ydb/ydb/core/tx/schemeshard/schemeshard_import.cpp:115:44: error: no member named 'IsDone' in 'NKikimr::NSchemeShard::TImportInfo'
  115 |             import.SetProgress(importInfo->IsDone()
      |  
```

By the way, what difficulties do you encounter when trying to build it locally? Why it doesn't work for you?

NeimadTL replied:
Let's just say that the computer that I have at work is not the best I've ever had. Most of the time while trying to build, it shuts down the terminal unfortunately. Thanks for the review.

pixcc replied:
There are discord and telegram chat, you can find the links at the bottom of the following site:

https://ydb.tech/

NeimadTL replied:
My eyes seen all of them except the Discord. Thanks a lot !

NeimadTL committed:
Fix typo

A parenthesis was missing.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import.cpp:
```diff
@@ -112,7 +112,7 @@ void TSchemeShard::FromXxportInfo(NKikimrImport::TImport& import, const TImportI
                 FillItemProgress(this, importInfo, itemIdx, *import.AddItemsProgress());
             }
-            import.SetProgress(importInfo->IsDone()
+            import.SetProgress(importInfo->IsDone())
```
Actually, that parenthesis is unnecessary, because there is no such method `IsDone` in `TImportInfo` struct. You need to add it like this:

```
    bool IsDone() const {
        return State == EState::Done;
    }
 ```

NeimadTL replied:
ok, thanks.

Curious though, what is [this](https://github.com/ydb-platform/ydb/blob/fb2c723ca8e9b4b49f492db5625af6b897327de9/ydb/core/tx/schemeshard/schemeshard_info_types.h#L2844) ? a completely different method definition as it take it takes a parameter ? a contract which forces you to have there ? or just unnecessary code ?

NeimadTL replied:
Understood, thanks.

NeimadTL replied:
Thank you.

NeimadTL committed:
Add `isDone` method

The implementation of the method was missing so it's been added.

NeimadTL committed:
Remove unnecessary parenthesis

This parenthesis was added by mistake misreading the code.

pixcc commented:
@NeimadTL Unfortunately the test you created failed again, could you please try to debug it?

NeimadTL commented:
> @NeimadTL Unfortunately the test you created failed again, could you please try to debug it?

Yes, will try to.

pixcc commented:
> > @NeimadTL Unfortunately the test you created failed again, could you please try to debug it?
> 
> Yes, will try to.

Hey, how it is going? Need help?

NeimadTL commented:
> > > @NeimadTL Unfortunately the test you created failed again, could you please try to debug it?
> > 
> > 
> > Yes, will try to.
> 
> @NeimadTL Hey, how it is going? Need help?

Hey @pixcc , thanks for reaching out. Been sick since last Monday so I haven't been programming. is the best way to get you live (instant messaging) Discord ? Will comme to this work soon.

pixcc commented:
> > > > @NeimadTL Unfortunately the test you created failed again, could you please try to debug it?
> > > 
> > > 
> > > Yes, will try to.
> > 
> > 
> > @NeimadTL Hey, how it is going? Need help?
> 
> Hey @pixcc , thanks for reaching out. Been sick since last Monday so I haven't been programming. is the best way to get you live (instant messaging) Discord ? Will come back to this work soon.

Hey @NeimadTL , what's up? Need any help? I have discord and telegram, what is the best way for you to communicate?
